### PR TITLE
Issue 46: Add Hashicorp Vault to the All-in-one management hub

### PR DESCRIPTION
**Pull request by naphelps**
References: #46
Depends on: https://github.com/open-horizon/vault-exchange-auth/pull/29

Changes:
- Adds Vault service to Docker environment.
- Adds Vault config option to `agbot-tmpl.json`
- Adds Vault instance setup and configuration to `deploy-mgmt-hub.sh`
- Adds `vault-tmpl.json` for base Vault configuration.

Does not support HTTPS at this time.

### CODE REVIEW

**Review by bmpotter — CHANGES_REQUESTED** [review]

File: mgmt-hub/deploy-mgmt-hub.sh [hunk 1]
```diff
@@ -210,6 +210,25 @@ export EXCHANGE_INTERNAL_RETRIES=${EXCHANGE_INTERNAL_RETRIES:-12}   # the maximu
 export EXCHANGE_INTERNAL_INTERVAL=${EXCHANGE_INTERNAL_INTERVAL:-5}   # the number of seconds to wait between attempts to connect to the exchange during startup
 # Note: in this environment, we are not supporting letting them specify their own owner key pair (only using the built-in sample key pair)
 
+export VAULT_AUTH_PLUGIN_EXCHANGE=openhorizon-exchange
+export VAULT_PORT=${VAULT_PORT:-8200}
+export VAULT_DEV_LISTEN_ADDRESS=${VAULT_DEV_LISTEN_ADDRESS:-0.0.0.0:${VAULT_PORT}}
```
[the remark this answers is not in the document]
Comment (naphelps) [reply mgmt-hub/deploy-mgmt-hub.sh:215]: This is for vault inside the container. `VAULT_DEV_LISTEN_ADDRESS` is needed for the docker image.

File: mgmt-hub/deploy-mgmt-hub.sh [hunk 2]
```diff
@@ -618,6 +642,10 @@ else   # ubuntu and redhat
     export VOLUME_MODE=ro
 fi
 
+export VAULT_INSTANCE_DIR=${ETC}/vault/file  # Persistent volume for Vault instance
+export VAULT_KEYS_DIR=${ETC}/vault/keys
+VAULT_KEYS_FILE=$VAULT_KEYS_DIR/keys.json
```
[the remark this answers is not in the document]
Comment (naphelps) [reply mgmt-hub/deploy-mgmt-hub.sh:647]: It does. Line 732: `rm -dfr ${ETC}/vault`

File: mgmt-hub/deploy-mgmt-hub.sh [hunk 3]
```diff
@@ -699,15 +727,28 @@ if [[ -n "$STOP" ]]; then
...
     fi
     exit
 fi
 
+# Required directories for Vault.
+if [[ ! -d ${VAULT_INSTANCE_DIR} ]]; then
+    mkdir -p ${VAULT_INSTANCE_DIR}
+    chown -R 1001 ${VAULT_INSTANCE_DIR}
+fi
+if [[ ! -d ${VAULT_KEYS_DIR} ]]; then
```
[the remark this answers is not in the document]
Comment (naphelps) [reply mgmt-hub/deploy-mgmt-hub.sh:748]: Fixed.

File: mgmt-hub/deploy-mgmt-hub.sh [hunk 4]
```diff
@@ -972,10 +1015,129 @@ else
     chkHttp $? $httpCode 201 "changing pw of /orgs/$EXCHANGE_SYSTEM_ORG/users/admin" $CURL_ERROR_FILE $CURL_OUTPUT_FILE
 fi
 
-# Create or update the agbot in the system org, and configure it with the pattern and deployment policy orgs
-if [[ $(exchangeGet $HZN_EXCHANGE_URL/orgs/$EXCHANGE_SYSTEM_ORG/agbots/$AGBOT_ID) == 200 ]]; then
-    restartAgbot='true'   # we may be changing its token, so need to restart it. (If there is initially no agbot resource, the agbot will just wait until it appears)
+printf "${CYAN}------- Creating a Vault instance and preforming all setup and configuration operations ...${NC}\n"
+vaultAuthMethodCheck() {
```
[the remark this answers is not in the document]
Comment (naphelps) [reply mgmt-hub/deploy-mgmt-hub.sh:1019]: Fixed.

File: mgmt-hub/deploy-mgmt-hub.sh [hunk 5]
```diff
@@ -972,10 +1015,129 @@ else
...
+    vaultEnableAuthMethod
+}
+
+vaultVaildation() {
+    echo Found a Vault instance.
+    # Regenerated root user's token
+    if [[ -z $VAULT_ROOT_TOKEN ]]; then
+        VAULT_ROOT_TOKEN=$(cat $VAULT_KEYS_FILE | jq -r '.root_token')
+    elif [[ -n $VAULT_ROOT_TOKEN ]] && [[ $VAULT_ROOT_TOKEN != $(cat $VAULT_KEYS_FILE | jq -r '.root_token') ]]; then
+        jq -a $VAULT_ROOT_TOKEN '.root_token=$VAULT_ROOT_TOKEN' < $VAULT_KEYS_FILE > $VAULT_KEYS_FILE
```
[the remark this answers is not in the document]
Comment (naphelps) [reply mgmt-hub/deploy-mgmt-hub.sh:1084]: Made these into a todo for the time being. Attempted to regenerate the root token and vault rejected the new token I was given.

File: mgmt-hub/deploy-mgmt-hub.sh [hunk 6]
```diff
@@ -972,10 +1015,129 @@ else
...
+    vaultServiceCheck
+    if [[ $(cat $VAULT_STATUS_FILE | jq '.initialized') == false ]]; then
+        vaultInitialize
+    else
+        vaultVaildation
+    fi
+
+    # Cannot read custom configuration keys/values. Assume either its never been set, or it has changed every time.
+    echo Configuring auth method $VAULT_AUTH_PLUGIN_EXCHANGE for use with the Exchange...
+    docker-compose exec -e VAULT_TOKEN=$VAULT_ROOT_TOKEN vault vault write -address=$HZN_TRANSPORT://0.0.0.0:8200 auth/openhorizon/config url=$HZN_TRANSPORT://exchange-api:8080/v1 token=$VAULT_ROOT_TOKEN
```
[the remark this answers is not in the document]
Comment (naphelps) [reply mgmt-hub/deploy-mgmt-hub.sh:1132]: Fixed.

File: mgmt-hub/deploy-mgmt-hub.sh [hunk 7]
```diff
@@ -618,6 +642,10 @@ else   # ubuntu and redhat
     export VOLUME_MODE=ro
 fi
 
+export VAULT_INSTANCE_DIR=${ETC}/vault/file  # Persistent volume for Vault instance
```
[the remark this answers is not in the document]
Comment (naphelps) [reply mgmt-hub/deploy-mgmt-hub.sh:645]: Fixed.

File: mgmt-hub/deploy-mgmt-hub.sh (hunk 1 above)
[the remark this answers is not in the document]
Comment (bmpotter) [reply mgmt-hub/deploy-mgmt-hub.sh:215]: But it is also used in the `ports` section in `docker-compose.yml`, which means it is exposing that port to the host with the IP in `VAULT_DEV_LISTEN_ADDRESS`, not the IP in `HZN_LISTEN_IP`

File: mgmt-hub/deploy-mgmt-hub.sh (hunk 2 above)
[the remark this answers is not in the document]
Comment (bmpotter) [reply mgmt-hub/deploy-mgmt-hub.sh]: It looks like you commented out all of the lines that create and remove this file, but `vaultInitialize` still tries to use it.

**Review by bmpotter — APPROVED** [review]

**Pull request merged**